contracts/Presale.sol

```diff
@@ -0,0 +1,227 @@
+        _sync($currentRoundIndex, address(0), amountUSD, _msgSender());
+    }
+
+    function purchase(address account) public payable override whenNotPaused {
+        uint256 amountUSD = ethToUsd(msg.value);
+        _sync($currentRoundIndex, address(0), amountUSD, account);
+    }
+
+    function purchaseUSDC(uint256 amount) external override whenNotPaused {
+        address sender = _msgSender();
+
+        IERC20(USDC).transferFrom(sender, address(this), amount);
+
+        uint256 amountScaled = amount * USDC_TO_WEI_PRECISION;
+        _sync($currentRoundIndex, USDC, amountScaled, sender);
+    }
+
+    function purchaseDAI(uint256 amount) external override whenNotPaused {
+        address sender = _msgSender();
+
+        IERC20(DAI).transferFrom(sender, address(this), amount);
+        _sync($currentRoundIndex, DAI, amount, sender);
+    }
+
+    receive() external payable {
+        purchase();
+    }
+
+    function _sync(uint256 roundIndex, address asset, uint256 amountUSD, address account) private {
+        PresaleConfig memory _config = $config;
+        RoundConfig[] memory _rounds = $rounds;
+
+        uint256 _userTokensAllocated = $userTokensAllocated[account];
+        uint256 userAllocationRemaining = _config.maxUserAllocation - _userTokensAllocated;
+
+        uint256 remainingUSD = amountUSD;
+        uint256 roundAllocationRemaining;
+
+        for (uint256 i = roundIndex; i < _rounds.length; ++i) {
+            RoundConfig memory _round = _rounds[i];
+            roundAllocationRemaining = _round.tokensAllocated - ($raisedUSD[i] * _round.tokenPrice);
+
+            if (roundAllocationRemaining == 0) continue;
+
+            uint256 _tokensAllocated = remainingUSD / _round.tokenPrice;
+
+            if (_tokensAllocated > roundAllocationRemaining) {
+                _tokensAllocated = roundAllocationRemaining;
+            }
+            if (_tokensAllocated > userAllocationRemaining) {
+                _tokensAllocated = userAllocationRemaining;
+            }
+
+            if (_tokensAllocated > 0) {
+                uint256 tokensCostUSD = _tokensAllocated * _round.tokenPrice;
+                remainingUSD -= tokensCostUSD;
+
+                roundAllocationRemaining -= _tokensAllocated;
+                userAllocationRemaining -= _tokensAllocated;
+
+                if (asset == address(0)) {
+                    _deposit(i, asset, tokensCostUSD / ethPrice(), tokensCostUSD, account);
+                } else {
+                    _deposit(i, asset, tokensCostUSD, tokensCostUSD, account);
+                }
+            }
+        }
+
+        $userTokensAllocated[account] = _config.maxUserAllocation - userAllocationRemaining;
+
+        if (remainingUSD > 0) {
+            _refund(asset, remainingUSD, account);
+        }
+    }
+
+    function _deposit(uint256 roundIndex, address asset, uint256 amount, uint256 amountUSD, address account) private {
+        PresaleConfig memory _config = $config;
```

Comment:
better to pass this variable as a function argument, than to get it from storage again

Reply:
ended up not needing this one after updating `$raisedUSD` mapping to `$roundAllocated`